src/network/connection.js

```diff
@@ -94,10 +118,45 @@ module.exports = class Connection {
     this.shouldLogBuffers = env.KAFKAJS_DEBUG_PROTOCOL_BUFFERS === '1'
     this.shouldLogFetchBuffer =
       this.shouldLogBuffers && env.KAFKAJS_DEBUG_EXTENDED_PROTOCOL_BUFFERS === '1'
+
+    this.authenticatedAt = null
+    this.sessionLifetime = Long.ZERO
+    this.supportAuthenticationProtocol = null
+
+    /**
+     * @private
+     * @returns {Promise}
+     */
+    this[PRIVATE.AUTHENTICATE] = sharedPromiseTo(async () => {
+      if (this.sasl && !this.isAuthenticated()) {
+        const authenticator = new SASLAuthenticator(
+          this,
+          this.logger,
+          this.versions,
+          this.supportAuthenticationProtocol
+        )
+
+        await authenticator.authenticate()
+        this.authenticatedAt = process.hrtime()
+        this.sessionLifetime = Long.fromValue(authenticator.sessionLifetime)
+      }
+    })
+  }
+
+  getSupportAuthenticationProtocol() {
+    return this.supportAuthenticationProtocol
+  }
+
+  setSupportAuthenticationProtocol(isSupported) {
+    this.supportAuthenticationProtocol = isSupported
+  }
+
+  setVersions(versions) {
+    this.versions = versions
   }
 
-  get connected() {
-    return CONNECTED_STATUS.includes(this.connectionStatus)
+  isConnected() {
+    return this.connectionStatus === CONNECTION_STATUS.CONNECTED
```

Comment:
I think the reason we has `isConnected() === true` when the connection status is `disconnecting` is to allow inflight requests to drain while not accepting new requests. Is this handled differently somehow now to make this not necessary?